File: packages/flow_widget_ios/ios/Classes/FlowWidgetPlugin.swift
import Flutter
import UIKit

public class FlowWidgetPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private static let methodChannelName = "dev.flow_widget/methods"
    private static let eventChannelName = "dev.flow_widget/events"

    private var storage: FlowWidgetStorage?
    private var imageStore: FlowWidgetImageStore?
    private var updater: FlowWidgetUpdater?
    private var liveActivityManager: FlowWidgetLiveActivityManager?
    private var eventSink: FlutterEventSink?
    private var initialized = false
    private var debugLogging = false
    private var appGroupId: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlowWidgetPlugin()
        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            switch call.method {
            case "initialize":
                try handleInitialize(call.arguments, result: result)
            case "dispose":
                disposeNative()
                result(nil)
            case "getCapabilities":
                result(FlowWidgetCapabilitiesProvider.capabilities())
            case "getPlatformType":
                result("ios")
            case "getInstalledWidgets":
                try ensureInitialized(result: result) {
                    result(self.requireUpdater().getInstalledWidgets())
                }
            case "saveData":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let key = args["key"] as? String,
                          let wire = args["value"] as? [String: Any] else {
                        return badArgs(result, message: "saveData requires key and value")
                    }
                    try self.requireStorage().saveData(
                        key: key,
                        wire: wire,
                        groupId: args["groupId"] as? String
                    )
                    result(nil)
                }
            case "saveBatch":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let entries = args["entries"] as? [[String: Any]] else {
                        return badArgs(result, message: "saveBatch requires entries")
                    }
                    try self.requireStorage().saveBatch(
                        entries: entries,
                        groupId: args["groupId"] as? String
                    )
                    result(nil)
                }
            case "getData":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let key = args["key"] as? String else {
                        return badArgs(result, message: "getData requires key")
                    }
                    let value = try self.requireStorage().getData(
                        key: key,
                        groupId: args["groupId"] as? String
                    )
                    result(value)
                }
            case "getAllData":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any] else {
                        return badArgs(result, message: "getAllData requires arguments map")
                    }
                    let allData = try self.requireStorage().getAllData(
                        prefix: args["prefix"] as? String,
                        groupId: args["groupId"] as? String
                    )
                    result(allData)
                }
            case "removeData":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let key = args["key"] as? String else {
                        return badArgs(result, message: "removeData requires key")
                    }
                    self.requireStorage().removeData(
                        key: key,
                        groupId: args["groupId"] as? String
                    )
                    result(nil)
                }
            case "clearData":
                try ensureInitialized(result: result) {
                    let groupId = (call.arguments as? [String: Any])?["groupId"] as? String
                    self.requireStorage().clearData(groupId: groupId)
                    result(nil)
                }
            case "saveImage":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any] else {
                        return badArgs(result, message: "saveImage requires arguments")
                    }
                    let path = try self.requireImageStore().saveImage(args: args)
                    result(path)
                }
            case "removeImage":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let key = args["key"] as? String else {
                        return badArgs(result, message: "removeImage requires key")
                    }
                    self.requireImageStore().removeImage(key: key)
                    result(nil)
                }
            case "update":
                try ensureInitialized(result: result) {
                    guard let request = call.arguments as? [String: Any] else {
                        return badArgs(result, message: "update requires request")
                    }
                    try self.requireUpdater().update(request: request)
                    result(nil)
                }
            case "updateMany":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let requests = args["requests"] as? [[String: Any]] else {
                        return badArgs(result, message: "updateMany requires requests")
                    }
                    try self.requireUpdater().updateMany(requests: requests)
                    result(nil)
                }
            case "updateAll":
                try ensureInitialized(result: result) {
                    self.requireUpdater().updateAll()
                    result(nil)
                }
            case "setTimeline":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let widgetId = args["widgetId"] as? [String: Any],
                          let entries = args["entries"] as? [[String: Any]] else {
                        return badArgs(result, message: "setTimeline requires widgetId and entries")
                    }
                    try self.requireUpdater().setTimeline(widgetId: widgetId, entries: entries)
                    result(nil)
                }
            case "registerConfig":
                try ensureInitialized(result: result) {
                    guard let config = call.arguments as? [String: Any] else {
                        return badArgs(result, message: "registerConfig requires config")
                    }
                    try self.requireStorage().registerConfig(config)
                    result(nil)
                }
            case "requestPinWidget":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let name = args["name"] as? String else {
                        return badArgs(result, message: "requestPinWidget requires name")
                    }
                    let pinned = try self.requireUpdater().requestPinWidget(
                        name: name,
                        initialData: args["initialData"] as? [String: Any]
                    )
                    result(pinned)
                }
            case "startLiveActivity":
                try ensureInitialized(result: result) {
                    guard let config = call.arguments as? [String: Any] else {
                        return badArgs(result, message: "startLiveActivity requires config")
                    }
                    let activityId = try self.requireLiveActivityManager().start(config: config)
                    result(activityId)
                }
            case "updateLiveActivity":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let activityId = args["activityId"] as? String,
                          let data = args["data"] as? [String: Any] else {
                        return badArgs(result, message: "updateLiveActivity requires activityId and data")
                    }
                    try self.requireLiveActivityManager().update(activityId: activityId, data: data)
                    result(nil)
                }
            case "endLiveActivity":
                try ensureInitialized(result: result) {
                    guard let args = call.arguments as? [String: Any],
                          let activityId = args["activityId"] as? String else {
                        return badArgs(result, message: "endLiveActivity requires activityId")
                    }
                    try self.requireLiveActivityManager().end(
                        activityId: activityId,
                        finalData: args["finalData"] as? [String: Any],
                        dismissalDate: args["dismissalDate"] as? Int
                    )
                    result(nil)
                }
            case "getActiveLiveActivities":
                try ensureInitialized(result: result) {
                    result(self.requireLiveActivityManager().activeActivities())
                }
            default:
                result(FlutterMethodNotImplemented)
            }
        } catch let error as FlowWidgetLiveActivityError {
            if case .unsupported = error {
                unsupported(result, message: error.localizedDescription)
            } else {
                result(FlutterError(code: "live_activity_error", message: error.localizedDescription, details: nil))
            }
        } catch let error as FlowWidgetCodecError {
            badArgs(result, message: error.localizedDescription)
        } catch {
            log("Method \(call.method) failed: \(error.localizedDescription)")
            result(FlutterError(code: "error", message: error.localizedDescription, details: nil))
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func handleInitialize(_ arguments: Any?, result: @escaping FlutterResult) throws {
        let args = arguments as? [String: Any] ?? [:]
        appGroupId = args["appGroupId"] as? String
        let maxCache = (args["imageCacheMaxBytes"] as? NSNumber)?.intValue ?? 20 * 1024 * 1024
        debugLogging = (args["enableDebugLogging"] as? Bool) ?? false

        let suiteStorage = FlowWidgetStorage(suiteName: appGroupId)
        storage = suiteStorage
        let container = appGroupId.flatMap { FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: $0) }
        imageStore = FlowWidgetImageStore(containerURL: container, maxCacheBytes: maxCache)
        updater = FlowWidgetUpdater(storage: suiteStorage)
        liveActivityManager = FlowWidgetLiveActivityManager(storage: suiteStorage)
        liveActivityManager?.setEventEmitter { [weak self] event in
            DispatchQueue.main.async {
                self?.eventSink?(event)
            }
        }
        initialized = true
        log("Initialized with appGroupId=\(appGroupId ?? "standard")")
        result(nil)
    }

    private func disposeNative() {
        imageStore?.clear()
        storage = nil
        imageStore = nil
        updater = nil
        liveActivityManager = nil
        initialized = false
    }

    private func ensureInitialized(result: @escaping FlutterResult, block: () throws -> Void) throws {
        guard initialized else {
            result(FlutterError(code: "not_initialized", message: "Call initialize() before other methods", details: nil))
            return
        }
        try block()
    }

    private func badArgs(_ result: @escaping FlutterResult, message: String) {
        result(FlutterError(code: "bad_args", message: message, details: nil))
    }

    private func unsupported(_ result: @escaping FlutterResult, message: String) {
        result(FlutterError(code: "unsupported", message: message, details: nil))
    }

    private func requireStorage() -> FlowWidgetStorage {
        guard let storage else {
            preconditionFailure("Storage not initialized")
        }
        return storage
    }

    private func requireImageStore() -> FlowWidgetImageStore {
        guard let imageStore else {
            preconditionFailure("Image store not initialized")
        }
        return imageStore
    }

    private func requireUpdater() -> FlowWidgetUpdater {
        guard let updater else {
            preconditionFailure("Updater not initialized")
        }
        return updater
    }

    private func requireLiveActivityManager() -> FlowWidgetLiveActivityManager {
        guard let liveActivityManager else {
            preconditionFailure("Live activity manager not initialized")
        }
        return liveActivityManager
    }

    private func log(_ message: String) {
        if debugLogging {
            NSLog("[FlowWidgetPlugin] %@", message)
        }
    }
}

File: packages/flow_widget_android/android/src/main/kotlin/dev/flowwidget/android/FlowWidgetPlugin.kt
package dev.flowwidget.android

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** Android implementation of the flow_widget federated plugin. */
class FlowWidgetPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    companion object {
        private const val METHOD_CHANNEL = "dev.flow_widget/methods"
        private const val EVENT_CHANNEL = "dev.flow_widget/events"
        private const val UNSUPPORTED = "unsupported"
    }

    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private var storage: FlowWidgetStorage? = null
    private var imageStore: FlowWidgetImageStore? = null
    private var updater: FlowWidgetUpdater? = null
    private var eventSink: EventChannel.EventSink? = null
    private var initialized = false
    private var debugLogging = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        disposeNative()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "initialize" -> handleInitialize(call, result)
                "dispose" -> handleDispose(result)
                "getCapabilities" -> result.success(FlowWidgetCapabilitiesProvider.getCapabilities())
                "getPlatformType" -> result.success("android")
                "getInstalledWidgets" -> ensureInitialized(result) {
                    result.success(requireUpdater().getInstalledWidgets())
                }
                "saveData" -> ensureInitialized(result) {
                    val key = call.argument<String>("key")
                        ?: return@ensureInitialized badArgs(result, "saveData requires key")
                    val value = call.argument<Map<String, Any?>>("value")
                        ?: return@ensureInitialized badArgs(result, "saveData requires value")
                    requireStorage().saveData(key, value, call.argument("groupId"))
                    result.success(null)
                }
                "saveBatch" -> ensureInitialized(result) {
                    val entries = call.argument<List<Map<String, Any?>>>("entries")
                        ?: return@ensureInitialized badArgs(result, "saveBatch requires entries")
                    requireStorage().saveBatch(entries, call.argument("groupId"))
                    result.success(null)
                }
                "getData" -> ensureInitialized(result) {
                    val key = call.argument<String>("key")
                        ?: return@ensureInitialized badArgs(result, "getData requires key")
                    result.success(requireStorage().getData(key, call.argument("groupId")))
                }
                "getAllData" -> ensureInitialized(result) {
                    val allData = requireStorage().getAllData(
                        prefix = call.argument("prefix"),
                        groupId = call.argument("groupId"),
                    )
                    result.success(allData)
                }
                "removeData" -> ensureInitialized(result) {
                    val key = call.argument<String>("key")
                        ?: return@ensureInitialized badArgs(result, "removeData requires key")
                    requireStorage().removeData(key, call.argument("groupId"))
                    result.success(null)
                }
                "clearData" -> ensureInitialized(result) {
                    requireStorage().clearData(call.argument("groupId"))
                    result.success(null)
                }
                "saveImage" -> ensureInitialized(result) {
                    val args = call.arguments as? Map<*, *>
                        ?: return@ensureInitialized badArgs(result, "saveImage requires arguments")
                    result.success(requireImageStore().saveImage(args))
                }
                "removeImage" -> ensureInitialized(result) {
                    val key = call.argument<String>("key")
                        ?: return@ensureInitialized badArgs(result, "removeImage requires key")
                    requireImageStore().removeImage(key)
                    result.success(null)
                }
                "update" -> ensureInitialized(result) {
                    val request = call.arguments as? Map<*, *>
                        ?: return@ensureInitialized badArgs(result, "update requires request")
                    requireUpdater().update(request)
                    result.success(null)
                }
                "updateMany" -> ensureInitialized(result) {
                    val requests = call.argument<List<Map<String, Any?>>>("requests")
                        ?: return@ensureInitialized badArgs(result, "updateMany requires requests")
                    requireUpdater().updateMany(requests)
                    result.success(null)
                }
                "updateAll" -> ensureInitialized(result) {
                    requireUpdater().updateAll()
                    result.success(null)
                }
                "setTimeline" -> ensureInitialized(result) {
                    val widgetId = call.argument<Map<String, Any?>>("widgetId")
                        ?: return@ensureInitialized badArgs(result, "setTimeline requires widgetId")
                    val entries = call.argument<List<Map<String, Any?>>>("entries")
                        ?: return@ensureInitialized badArgs(result, "setTimeline requires entries")
                    requireUpdater().setTimeline(widgetId, entries)
                    result.success(null)
                }
                "registerConfig" -> ensureInitialized(result) {
                    val config = call.arguments as? Map<*, *>
                        ?: return@ensureInitialized badArgs(result, "registerConfig requires config")
                    requireStorage().registerConfig(config)
                    result.success(null)
                }
                "requestPinWidget" -> ensureInitialized(result) {
                    val name = call.argument<String>("name")
                        ?: return@ensureInitialized badArgs(result, "requestPinWidget requires name")
                    val initialData = call.argument<Map<String, Any?>>("initialData")
                    val pinned = requireUpdater().requestPinWidget(name, initialData)
                    result.success(pinned)
                }
                "startLiveActivity",
                "updateLiveActivity",
                "endLiveActivity",
                "getActiveLiveActivities" -> unsupportedLiveActivity(result)
                else -> result.notImplemented()
            }
        } catch (error: IllegalArgumentException) {
            badArgs(result, error.message ?: "Invalid arguments")
        } catch (error: IllegalStateException) {
            result.error("error", error.message, null)
        } catch (error: Exception) {
            log("Method ${call.method} failed: ${error.message}")
            result.error("error", error.message, null)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun emitEvent(event: Map<String, Any?>) {
        val sink = eventSink ?: return
        Handler(Looper.getMainLooper()).post { sink.success(event) }
    }

    private fun handleInitialize(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *> ?: emptyMap<Any?, Any?>()
        val androidPrefs = args["androidPrefs"] as? String
        val appGroupId = args["appGroupId"] as? String
        // Android prefs are independent of iOS/macOS appGroupId. Never fall back
        // to appGroupId — that silently mismatches native FlowWidgetStorage.create.
        val prefsName = androidPrefs ?: FlowWidgetStorage.DEFAULT_PREFS_NAME
        val useGlance = args["useGlance"] as? Boolean ?: true
        val maxCache = (args["imageCacheMaxBytes"] as? Number)?.toLong() ?: 20L * 1024L * 1024L
        debugLogging = args["enableDebugLogging"] == true

        if (androidPrefs.isNullOrEmpty() && !appGroupId.isNullOrEmpty()) {
            android.util.Log.w(
                "FlowWidgetPlugin",
                "androidNamedSharedPreferences was null; using " +
                    "\"${FlowWidgetStorage.DEFAULT_PREFS_NAME}\". appGroupId " +
                    "(\"$appGroupId\") is iOS/macOS-only and is not used as the " +
                    "Android SharedPreferences name. Pass the same name to " +
                    "FlowWidgetStorage.create in your Glance/RemoteViews code.",
            )
        }

        storage = FlowWidgetStorage.create(context, prefsName)
        imageStore = FlowWidgetImageStore(context, maxCache)
        updater = FlowWidgetUpdater(context, requireStorage(), useGlance = useGlance)
        initialized = true
        log("Initialized with prefs=$prefsName useGlance=$useGlance")
        result.success(null)
    }

    private fun handleDispose(result: MethodChannel.Result) {
        disposeNative()
        result.success(null)
    }

    private fun disposeNative() {
        imageStore?.clear()
        storage = null
        imageStore = null
        updater = null
        initialized = false
    }

    private fun ensureInitialized(result: MethodChannel.Result, block: () -> Unit) {
        if (!initialized) {
            result.error("not_initialized", "Call initialize() before other methods", null)
            return
        }
        block()
    }

    private fun unsupportedLiveActivity(result: MethodChannel.Result) {
        result.error(
            UNSUPPORTED,
            "Live Activities are not supported on Android",
            null,
        )
    }

    private fun badArgs(result: MethodChannel.Result, message: String) {
        result.error("bad_args", message, null)
    }

    private fun requireStorage(): FlowWidgetStorage {
        return storage ?: throw IllegalStateException("Storage not initialized")
    }

    private fun requireImageStore(): FlowWidgetImageStore {
        return imageStore ?: throw IllegalStateException("Image store not initialized")
    }

    private fun requireUpdater(): FlowWidgetUpdater {
        return updater ?: throw IllegalStateException("Updater not initialized")
    }

    private fun log(message: String) {
        if (debugLogging) {
            android.util.Log.d("FlowWidgetPlugin", message)
        }
    }
}
